### ISSUE TITLE
arrowkeys_normal.patch shouldn't exit insert mode

### ISSUE THREAD

**Issue by OldWorldOrdr** [kyx0r/nextvi#62]
It's annoying to have to reenter insert mode every time I use the arrow keys

**Comment by OldWorldOrdr** [comment]
when in insert mode, using arrow keys enters normal mode

**Comment by kyx0r** [comment]
Its possible, kind of tricky. You could make a macro that does this while in insert mode (which is way easier but has drawbacks), or bring out the whole state in vi_insmov

**Comment by kyx0r** [comment]
https://github.com/kyx0r/nextvi/blob/patches/arrowkeys_insert.patch

Try, this approach has side effects, but is the simplest way to get it to work like you described.

-Kyryl

**Comment by OldWorldOrdr** [comment]
on the right track but very buggy

breaks syntax highlighting on comments when moving up and down

moving up and down also moves the cursor to the left by one each move

the first keypress after exiting insert mode does nothing

if you open a file, enter insert mode, then exit insert mode and try to quit with :q it will say that the buffer was modified, even if nothing was changed

![image](https://github.com/kyx0r/nextvi/assets/57970114/20244211-540f-4eac-974b-23cd7783b6ed)

**Comment by kyx0r** [comment]
That's to be expected.

1. Syntax will be broken because it re-enters insert.
2. That's because vi_col is updated with syntax
3. Because its not a buffered input read, 1 character at a time. Use ^c for now
4. Old problem, in fact insert will always do this even if nothing was changed. Ideally this should be fixed in master but I'm not sure about the best way.

**Comment by kyx0r** [comment]
Ok, issue 4. should be fixed now. Hopefully without regressions in somewhere else. That one was annoying me for quite a while.

**Comment by OldWorldOrdr** [comment]
both 2 and 4 are fixed

**Comment by OldWorldOrdr** [comment]
Seems to me like this might be a bit too complex to get to a usable state without a lot more work, the issues with syntax highlighting and the first keypress after leaving insert being ignored are still problems, I would help out but I'm not nearly good enough at C to do something this complex, should I close this issue?

**Comment by kyx0r** [comment]
No, don't close this. Those problems can be addressed and fixed, I just haven't got around to doing so yet. The patch will become a bit more involved to rule out these edge cases but not too bad.

**Comment by kyx0r** [comment]
@OldWorldOrdr Try the new version.

**Comment by OldWorldOrdr** [comment]
the issue with the first keypress being ignored is fixed, syntax highlighting is *almost* fixed, it still breaks whatever line is highlighted when entering or exiting insert mode

**Comment by kyx0r** [comment]
I don't think its possible to fix syntax here without some major architecture change. When in insert mode the line is redrawn, and visually positioned. It can't know any multiline highlight state either.

**Comment by kyx0r** [comment]
Well, it could be done if we force redraw the whole screen every time, but that is less ideal and hurts performance. But it would make the patch simpler too.

**Comment by OldWorldOrdr** [comment]
i have some super low power devices i can test on, if you try it ill let you know if the performance hit is noticeable.

**Comment by kyx0r** [comment]
What's more bothersome to me is that the col position isn't preserved. Because i command has side effects, for example when exiting insert its supposed to go back 1 character. So we have to negate that with l command, but that causes the col to get updated which is why it doesn't behave the same way regular jk motions work.

The core could be changed to not redraw the line on entering insert mode, but thats also kinda ewk.

**Comment by kyx0r** [comment]
We need an insert mode that doesn't have side effects. At same time I don't want to break backwards compat with traditional vi i command.

**Comment by kyx0r** [comment]
a command looks like could work better, and also will enable us to move the cursor to the end of the line. i.e possible to append at the very end.

**Comment by kyx0r** [comment]
@OldWorldOrdr Now try the new version. I hope you come to appreciate the col position improvements.

**Comment by kyx0r** [comment]
Added a check for that.
I think the core needs one more change, that is prevent line redraw on initial insert enter. Its a bit tricky to do and I have to see if adding "dead ifs" statements in loops is worthwhile.

**Comment by OldWorldOrdr** [comment]
> it still breaks whatever line is highlighted when entering or exiting insert mode

Took me way too long to realize this also happens without the patch, and also happens if you change lines in insert mode then start adding text to a different line, I think we can consider this done, since the only bugs still present are also bugs in master

Should I close this and open a new issue for the syntax highlighting issues?

**Comment by kyx0r** [comment]
Yeah, the patch is probably done. 

Keep in mind that the reason these bugs exist with multiline syntax is because of a simple architecture of how its implemented. There aren't any data structures involved in tracking it and the thing is like 20 lines of c logic. Also its probably makes more sense to not attempt to redraw the multiline syntax when in insert, as it helps the user see what he is editing.
The only thing I would change is not redrawing it pre-emptively, only after user starts to make changes.

**Comment by kyx0r** [comment]
Hmm, this is a lot easier said than done. Apparently this is tightly coupled with the prompt and not only that, some operations rely on this redraw. Plus in the context of nested loops and functions its difficult to make this work without resorting to global variables.

I leave this be for now, maybe later better ideas will come.
